File: src/main/java/io/nightfis/createstellarexoduscore/network/AutoAimTurretFilterSyncPacket.java
package io.nightfis.createstellarexoduscore.network;

import io.nightfis.createstellarexoduscore.client.gui.screen.AutoAimTurretScreen;
import io.nightfis.createstellarexoduscore.filter.TargetFilter;
import it.unimi.dsi.fastutil.Pair;
import net.minecraft.client.Minecraft;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraftforge.network.NetworkEvent;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Supplier;

public class AutoAimTurretFilterSyncPacket {

    private final List<Pair<TargetFilter, String>> filters;

    public AutoAimTurretFilterSyncPacket(List<Pair<TargetFilter, String>> filters) {
        this.filters = new ArrayList<>(filters);
    }

    public static void encode(AutoAimTurretFilterSyncPacket msg, FriendlyByteBuf buf) {
        buf.writeCollection(msg.filters.stream().map(it -> it.left().getId()).toList(), FriendlyByteBuf::writeUtf);
        buf.writeCollection(msg.filters.stream().map(Pair::right).toList(), FriendlyByteBuf::writeUtf);
    }

    public static AutoAimTurretFilterSyncPacket decode(FriendlyByteBuf buf) {
        List<String> ids = buf.readList(FriendlyByteBuf::readUtf);
        List<String> args = buf.readList(FriendlyByteBuf::readUtf);
        ArrayList<Pair<TargetFilter, String>> filters = new ArrayList<>();
        for (int i = 0; i < args.size() && i < ids.size(); i++) {
            filters.add(Pair.of(TargetFilter.getFilter(ids.get(i)), args.get(i)));
        }
        return new AutoAimTurretFilterSyncPacket(filters);
    }

    public static void handle(AutoAimTurretFilterSyncPacket msg, Supplier<NetworkEvent.Context> ctx) {
        ctx.get().enqueueWork(() -> {
            if (Minecraft.getInstance().screen instanceof AutoAimTurretScreen screen) {
                screen.handleSync(msg.filters);
            }
        });
        ctx.get().setPacketHandled(true);
    }
}

File: src/main/kotlin/io/nightfis/createstellarexoduscore/network/AutoAimTurretFilterSyncPacket.kt
package io.nightfis.createstellarexoduscore.network

import io.nightfis.createstellarexoduscore.client.gui.screen.AutoAimTurretScreen
import io.nightfis.createstellarexoduscore.filter.TargetFilter
import it.unimi.dsi.fastutil.Pair
import net.minecraft.client.Minecraft
import net.minecraft.network.FriendlyByteBuf
import net.minecraftforge.network.NetworkEvent
import java.util.ArrayList
import java.util.function.Supplier

class AutoAimTurretFilterSyncPacket(filters: List<Pair<TargetFilter, String>>) {

    val filters: List<Pair<TargetFilter, String>> = ArrayList(filters)

    companion object {
        fun encode(msg: AutoAimTurretFilterSyncPacket, buf: FriendlyByteBuf) {
            buf.writeCollection(msg.filters.map { it.left().getId() }, FriendlyByteBuf::writeUtf)
            buf.writeCollection(msg.filters.map { it.right() }, FriendlyByteBuf::writeUtf)
        }

        fun decode(buf: FriendlyByteBuf): AutoAimTurretFilterSyncPacket {
            val ids = buf.readList(FriendlyByteBuf::readUtf)
            val args = buf.readList(FriendlyByteBuf::readUtf)
            val filters = ArrayList<Pair<TargetFilter, String>>()
            val count = minOf(args.size, ids.size)
            for (i in 0 until count) {
                filters.add(Pair.of(TargetFilter.getFilter(ids[i]), args[i]))
            }
            return AutoAimTurretFilterSyncPacket(filters)
        }

        fun handle(msg: AutoAimTurretFilterSyncPacket, ctx: Supplier<NetworkEvent.Context>) {
            val context = ctx.get()
            context.enqueueWork {
                val screen = Minecraft.getInstance().screen
                if (screen is AutoAimTurretScreen) {
                    screen.handleSync(msg.filters)
                }
            }
            context.packetHandled = true
        }
    }
}
